### ISSUE TITLE
CodeAction testing infrastructure does not behave hte same as VS wrt EquivalenceKey

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#44553]
See https://github.com/dotnet/roslyn/pull/44549 as an example of an issue caused by this.
 
We have a code-fixer "Generate constructor" which can produce up to two actions to fix a diagnostic.  specifically:

1. Generate constructor
2. Generate constructor (without fields)

This feature has unit tests validating it works and that you should get both of these (and how each individually works).  

However, hte actual VS product only shows the first item due to the second having the same equivalence-key as the first (`null`) and being produced by the same fixer for the same diagnostic.

--

This demonstrates a test hole we have.  The tests are supposed to mimic the user behavior and validate what hte user sees and what happens when the user invokes particular options presented ot htem.  In this case, because the code fix service does not show the second fix, the test should act as though only one fix is made available and should throw for tests that try to invoke the second fix.

**Comment by CyrusNajmabadi** [comment]
Tagging @mavasani

**Comment by sharwell** [comment]
@CyrusNajmabadi Can you file this issue in dotnet/roslyn-sdk as well? The test harness here is legacy at this point and we don't want the test gap to reappear in the replacement.

**Comment by CyrusNajmabadi** [comment]
This needs to be bumped up in priority.  I've now found two features that regressed here because of some change in behavior that our test harness did not catch.  

Specifically, i believe in https://github.com/dotnet/roslyn/pull/41768 we changed the core behavior of the CodeFixService so that it now consider the fixes equivalent *even if* they have `null` `.EquivalenceKey`s.  This violates the docs for them this property which says:

```
        /// <summary>
        /// Two code actions are treated as equivalent if they have equal non-null <see cref="EquivalenceKey"/> values and were generated
        /// by the same <see cref="CodeFixProvider"/> or <see cref="CodeRefactoringProvider"/>.
        /// </summary>
```

This change in behavior may affect external customers as well.  My preference would be that we restore the original behavior that didn't dedupe actions with a `null` key.  However, if we do want to keep this new behavior, it's paramount we fix our testing harness to actually now represent this behavior so we can detect and fix immediately any other regressions we may have.
